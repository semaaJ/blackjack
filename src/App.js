import React, { useState } from 'react';
import Card from './components/Card/Card';
import Gamebar from './components/Gamebar/Gamebar';
import { dealRound, initialiseGame, playerBet, playerHit, playerStand, playerDoubleDown  } from './utils/BlackJack';

export const renderHouseCards = (cards, showSecond) => {
  return cards.length > 0 ? (
    <>
      { <Card rank={cards[0].rank} symbol={cards[0].symbol} colour={cards[0].colour} /> }
      { 
        cards.slice(1, cards.length).map((_, i) => showSecond ? 
          <Card rank={cards[i + 1].rank} symbol={cards[i + 1].symbol} clour={cards[i + 1].colour} /> :
          <div className="bg-slate-500 -ml-28 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2 sm:ml-0"  />
        )
      }
    </>
  ) : (
    <>
      <div className="bg-slate-500 -ml-28 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2 sm:ml-0"  />
      <div className="bg-slate-500 -ml-28 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2 sm:ml-0"  />
    </>
  )
}

export const renderPlayerCards = (cards) => {
  return cards.length > 0 ? ( 
    <>
      { cards.map(card => <Card rank={card.rank} symbol={card.symbol} colour={card.colour} />) }
    </>
  ) : (
    <>
        <div className="bg-slate-500 -ml-28 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2 sm:ml-0"  />
        <div className="bg-slate-500 -ml-28 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2 sm:ml-0"  />
    </>
  )
}

export const bet = (state, setState, amount) => setState(playerBet(state, amount));
export const hit = (state, setState) => setState(playerHit(state));
export const doubleDown = (state, setState) => setState(playerDoubleDown(state));
export const stand = (state, setState) => setState({ ...playerStand(state), showSecond: true });
export const deal = (state, setState) => setState(dealRound(state));
export const reset = (setState) => setState(initialiseGame());

function App() {
  const [state, setState] = useState(initialiseGame());
  return (
      <div className="w-screen overflow-x-hidden h-screen bg-green-800 flex flex-col items-center font-roboto">
        <div className="flex flex-col justify-evenly h-full max-w-7xl">
          <div className="flex translate-x-14 w-full justify-center md:transform-none">
            { renderHouseCards(state.house.cards, state.showSecond) }
          </div>

          <div className="flex translate-x-14 w-full justify-center md:transform-none">
              { renderPlayerCards(state.player.cards) }
          </div>

          <div className="flex items-center bg-white justify-center">
            <Gamebar 
              state={state} 
              setState={setState}
              bet={bet} 
              hit={hit} 
              doubleDown={doubleDown}
              stand={stand} 
              deal={deal}
              reset={reset}
            />
          </div>
        </div> 
    </div>  
  );
}

export default App;
