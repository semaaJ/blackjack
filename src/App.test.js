import React from 'react';
import App, { renderHouseCards, renderPlayerCards } from './App';
import renderer from 'react-test-renderer';

const cards = [
    { rank: "K", symbol: "K", colour:"text-rose-500"}, 
    { rank: "K", symbol: "K", colour:"text-rose-500"}
];

it("should render", () => {
    const tree = renderer.create(<App />).toJSON();    
    expect(tree).toMatchSnapshot();
});

it("should only show one dealer card on secondShow is false", () => {
    const tree = renderer.create(renderHouseCards(cards)).toJSON();    
    expect(tree).toMatchSnapshot();
});

it("should render all house cards on secondShow", () => {
    const tree = renderer.create(renderHouseCards(cards, true)).toJSON();    
    expect(tree).toMatchSnapshot();
});

it("should render player cards", () => {
    const tree = renderer.create(renderPlayerCards(cards)).toJSON();    
    expect(tree).toMatchSnapshot();
});