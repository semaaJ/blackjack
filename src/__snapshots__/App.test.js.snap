// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should only show one dealer card on secondShow is false 1`] = `
Array [
  <div
    className="bg-white px-3 w-40 h-60 shadow-2xl rounded-lg sm:m-2"
  >
    <div
      className="flex flex-col justify-center"
      style={
        Object {
          "backfaceVisibility": "hidden",
        }
      }
    >
      <div
        className=""
      >
        <div
          className="flex flex-col w-full items-start"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
        <div
          className="flex justify-center"
        >
          <h1
            className="text-7xl text-rose-500"
          >
            K
          </h1>
        </div>
        <div
          className="flex flex-col w-full items-end"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
  />,
  <div
    className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
  />,
]
`;

exports[`should render 1`] = `
<div
  className="w-screen overflow-x-hidden h-screen flex flex-col items-center bg-green-800 font-roboto"
>
  <div
    className="flex flex-col justify-evenly h-full max-w-7xl"
  >
    <div
      className="flex flex-col"
    >
      <div
        className="flex flex-wrap justify-between sm:justify-center"
      >
        <div
          className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
        />
        <div
          className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
        />
        <div
          className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
        />
      </div>
    </div>
    <div
      className="flex items-center justify-center"
    >
      <div
        className="flex flex-col bg-white shadow-xl rounded-2xl border-4  border-white sm: w-full"
      >
        <div
          className="text-center w-full mb-2 mt-2"
        >
          <h1
            className="text-gray-600 font-bold text-2xl"
          >
            DEAL CARDS
          </h1>
        </div>
        <div
          className="flex justify-between"
        >
          <button
            className="w-btn-small h-btn text-center text-gray-700 text-sm bg-white border-white 
 sm:w-full sm:font-semibold sm:text-lg hover:bg-gray-200 disabled:bg-gray-200"
            disabled={false}
            id="deal"
            onClick={[Function]}
          >
            <div
              className="flex items-center justify-center"
            >
              DEAL
              <svg
                aria-hidden="true"
                className="ml-2 h-5 w-5 text-gray-500"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M4 2a2 2 0 00-2 2v11a3 3 0 106 0V4a2 2 0 00-2-2H4zm1 14a1 1 0 100-2 1 1 0 000 2zm5-1.757l4.9-4.9a2 2 0 000-2.828L13.485 5.1a2 2 0 00-2.828 0L10 5.757v8.486zM16 18H9.071l6-6H16a2 2 0 012 2v2a2 2 0 01-2 2z"
                  fillRule="evenodd"
                />
              </svg>
            </div>
          </button>
          <button
            className="w-btn-small h-btn text-center text-gray-700 text-sm bg-white border-white 
 sm:w-full sm:font-semibold sm:text-lg hover:bg-gray-200 disabled:bg-gray-200"
            disabled={true}
            id="hit"
            onClick={[Function]}
          >
            <div
              className="flex items-center justify-center"
            >
              HIT
              <svg
                aria-hidden="true"
                className="ml-2 h-5 w-5 text-gray-500"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10 18a8 8 0 100-16 8 8 0 000 16zM9.555 7.168A1 1 0 008 8v4a1 1 0 001.555.832l3-2a1 1 0 000-1.664l-3-2z"
                  fillRule="evenodd"
                />
              </svg>
            </div>
          </button>
          <button
            className="w-btn-small h-btn text-center text-gray-700 text-sm bg-white border-white 
 sm:w-full sm:font-semibold sm:text-lg hover:bg-gray-200 disabled:bg-gray-200"
            disabled={true}
            id="doubleDown"
            onClick={[Function]}
          >
            <div
              className="flex items-center justify-center"
            >
              DOUBLE DOWN
              <h1
                className="ml-2 text-base text-gray-500"
              >
                x2
              </h1>
            </div>
          </button>
          <button
            className="w-btn-small h-btn text-center text-gray-700 text-sm bg-white border-white 
 sm:w-full sm:font-semibold sm:text-lg hover:bg-gray-200 disabled:bg-gray-200"
            disabled={true}
            id="stand"
            onClick={[Function]}
          >
            <div
              className="flex items-center justify-center"
            >
              STAND
              <svg
                aria-hidden="true"
                className="ml-2 h-5 w-5 text-gray-500"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 12a2 2 0 100-4 2 2 0 000 4z"
                />
                <path
                  clipRule="evenodd"
                  d="M.458 10C1.732 5.943 5.522 3 10 3s8.268 2.943 9.542 7c-1.274 4.057-5.064 7-9.542 7S1.732 14.057.458 10zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
                  fillRule="evenodd"
                />
              </svg>
            </div>
          </button>
          <button
            className="w-btn-small h-btn text-center text-gray-700 text-sm bg-white border-white 
 sm:w-full sm:font-semibold sm:text-lg hover:bg-gray-200 disabled:bg-gray-200"
            disabled={true}
            id="restart"
            onClick={[Function]}
          >
            <div
              className="flex items-center justify-center"
            >
              RESTART
              <svg
                aria-hidden="true"
                className="ml-2 h-5 w-5 text-gray-500"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M4 2a1 1 0 011 1v2.101a7.002 7.002 0 0111.601 2.566 1 1 0 11-1.885.666A5.002 5.002 0 005.999 7H9a1 1 0 010 2H4a1 1 0 01-1-1V3a1 1 0 011-1zm.008 9.057a1 1 0 011.276.61A5.002 5.002 0 0014.001 13H11a1 1 0 110-2h5a1 1 0 011 1v5a1 1 0 11-2 0v-2.101a7.002 7.002 0 01-11.601-2.566 1 1 0 01.61-1.276z"
                  fillRule="evenodd"
                />
              </svg>
            </div>
          </button>
        </div>
        <div
          className="flex bg-gray-200 justify-center"
        >
          <h1
            className="bg-gray-200 text-gray-900 text-center text-lg mt-1"
          >
            SCORE: 
            0
          </h1>
          <h1
            className="bg-gray-200 ml-20 text-gray-900 text-center text-lg mt-1"
          >
            HOUSE: 
            ?
          </h1>
        </div>
        <div
          className="flex justify-evenly bg-gray-200 items-center rounded-bl-2xl rounded-br-2xl"
        >
          <div
            className="flex text-center flex-col"
          >
            <h1
              className="text-gray-700 text-lg sm:text-2xl sm:font-bold"
            >
              BET
            </h1>
            <h1
              className="text-gray-700 text-lg sm:text-xl sm:font-bold"
            >
              $
              0
            </h1>
          </div>
          <div
            className="flex justify-center py-4"
          >
            <img
              alt="Chip 5 value"
              className="sm:ml-4 sm:mr-4 w-12 h-12 sm:w-16 sm:h-16 grayscale"
              onClick={[Function]}
              src="https://s3-us-west-2.amazonaws.com/s.cdpn.io/1151732/chip-5.png"
            />
            <img
              alt="Chip 10 value"
              className="sm:ml-4 sm:mr-4 w-12 h-12 sm:w-16 sm:h-16 grayscale"
              onClick={[Function]}
              src="https://s3-us-west-2.amazonaws.com/s.cdpn.io/1151732/chip-7.png"
            />
            <img
              alt="Chip 15 value"
              className="sm:ml-4 sm:mr-4 w-12 h-12 sm:w-16 sm:h-16 grayscale"
              onClick={[Function]}
              src="https://s3-us-west-2.amazonaws.com/s.cdpn.io/1151732/chip-4.png"
            />
            <img
              alt="Chip 50 value"
              className="sm:ml-4 sm:mr-4 w-12 h-12 sm:w-16 sm:h-16 grayscale"
              onClick={[Function]}
              src="https://s3-us-west-2.amazonaws.com/s.cdpn.io/1151732/chip-3.png"
            />
          </div>
          <div
            className="flex text-center flex-col"
          >
            <h1
              className="text-gray-700 text-lg sm:text-2xl sm:font-bold"
            >
              CHIPS
            </h1>
            <h1
              className="text-gray-700 text-lg sm:text-xl sm:font-bold"
            >
              $
              1000
            </h1>
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex flex-col"
    >
      <div
        className="flex flex-wrap justify-between sm:justify-center"
      >
        <div
          className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
        />
        <div
          className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
        />
        <div
          className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
        />
        <div
          className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
        />
        <div
          className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should render all house cards on secondShow 1`] = `
Array [
  <div
    className="bg-white px-3 w-40 h-60 shadow-2xl rounded-lg sm:m-2"
  >
    <div
      className="flex flex-col justify-center"
      style={
        Object {
          "backfaceVisibility": "hidden",
        }
      }
    >
      <div
        className=""
      >
        <div
          className="flex flex-col w-full items-start"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
        <div
          className="flex justify-center"
        >
          <h1
            className="text-7xl text-rose-500"
          >
            K
          </h1>
        </div>
        <div
          className="flex flex-col w-full items-end"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bg-white px-3 w-40 h-60 shadow-2xl rounded-lg sm:m-2"
  >
    <div
      className="flex flex-col justify-center"
      style={
        Object {
          "backfaceVisibility": "hidden",
        }
      }
    >
      <div
        className=""
      >
        <div
          className="flex flex-col w-full items-start"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl undefined"
            >
              K
            </h1>
            <h1
              className="text-4xl undefined"
            >
              K
            </h1>
          </div>
        </div>
        <div
          className="flex justify-center"
        >
          <h1
            className="text-7xl undefined"
          >
            K
          </h1>
        </div>
        <div
          className="flex flex-col w-full items-end"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl undefined"
            >
              K
            </h1>
            <h1
              className="text-4xl undefined"
            >
              K
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bg-white px-3 w-40 h-60 shadow-2xl rounded-lg sm:m-2"
  >
    <div
      className="flex flex-col justify-center"
      style={
        Object {
          "backfaceVisibility": "hidden",
        }
      }
    >
      <div
        className=""
      >
        <div
          className="flex flex-col w-full items-start"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl undefined"
            >
              K
            </h1>
            <h1
              className="text-4xl undefined"
            >
              K
            </h1>
          </div>
        </div>
        <div
          className="flex justify-center"
        >
          <h1
            className="text-7xl undefined"
          >
            K
          </h1>
        </div>
        <div
          className="flex flex-col w-full items-end"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl undefined"
            >
              K
            </h1>
            <h1
              className="text-4xl undefined"
            >
              K
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`should render player cards 1`] = `
Array [
  <div
    className="bg-white px-3 w-40 h-60 shadow-2xl rounded-lg sm:m-2"
  >
    <div
      className="flex flex-col justify-center"
      style={
        Object {
          "backfaceVisibility": "hidden",
        }
      }
    >
      <div
        className=""
      >
        <div
          className="flex flex-col w-full items-start"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
        <div
          className="flex justify-center"
        >
          <h1
            className="text-7xl text-rose-500"
          >
            K
          </h1>
        </div>
        <div
          className="flex flex-col w-full items-end"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bg-white px-3 w-40 h-60 shadow-2xl rounded-lg sm:m-2"
  >
    <div
      className="flex flex-col justify-center"
      style={
        Object {
          "backfaceVisibility": "hidden",
        }
      }
    >
      <div
        className=""
      >
        <div
          className="flex flex-col w-full items-start"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
        <div
          className="flex justify-center"
        >
          <h1
            className="text-7xl text-rose-500"
          >
            K
          </h1>
        </div>
        <div
          className="flex flex-col w-full items-end"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bg-white px-3 w-40 h-60 shadow-2xl rounded-lg sm:m-2"
  >
    <div
      className="flex flex-col justify-center"
      style={
        Object {
          "backfaceVisibility": "hidden",
        }
      }
    >
      <div
        className=""
      >
        <div
          className="flex flex-col w-full items-start"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
        <div
          className="flex justify-center"
        >
          <h1
            className="text-7xl text-rose-500"
          >
            K
          </h1>
        </div>
        <div
          className="flex flex-col w-full items-end"
        >
          <div
            className="flex flex-col items-center"
          >
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
            <h1
              className="text-4xl text-rose-500"
            >
              K
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
  />,
  <div
    className="bg-slate-500 px-3 w-40 h-60 shadow-xl rounded-lg sm:m-2"
  />,
]
`;
